arrow-libs/core/arrow-autoclose/src/commonMain/kotlin/arrow/AutoCloseScope.kt

```diff
@@ -61,53 +62,47 @@ import kotlin.coroutines.cancellation.CancellationException
  * ```
  * <!--- KNIT example-autocloseable-02.kt -->
  */
-public inline fun <A> autoCloseScope(block: AutoCloseScope.() -> A): A {
-  val scope = DefaultAutoCloseScope()
-  return try {
-    block(scope)
-      .also { scope.close(null) }
-  } catch (e: CancellationException) {
-    scope.close(e) ?: throw e
+public inline fun <A> autoCloseScope(block: AutoCloseScope.() -> A): A = with(DefaultAutoCloseScope()) {
+  try {
+    block()
   } catch (e: Throwable) {
-    scope.close(e.throwIfFatal()) ?: throw e
-  }
+    closeAll(e.throwIfFatal())
+    error("Unreachable, closeAll should throw the exception passed to it. Please report this bug to the Arrow team.")
+  }.also { closeAll(null) }
 }
 
 public interface AutoCloseScope {
-  public fun <A> autoClose(
-    acquire: () -> A,
-    release: (A, Throwable?) -> Unit
-  ): A
+  public fun onClose(release: (Throwable?) -> Unit)
 
   @ExperimentalStdlibApi
   public fun <A : AutoCloseable> install(autoCloseable: A): A =
-    autoClose({ autoCloseable }) { a, _ -> a.close() }
+    autoCloseable.also { onClose { autoCloseable.close() } }
 }
 
-@PublishedApi
-internal class DefaultAutoCloseScope : AutoCloseScope {
-  private val finalizers = Atomic(emptyList<(Throwable?) -> Unit>())
+public inline fun <A> AutoCloseScope.autoClose(
```

Comment:
Idem, this introduces a source breaking change, to support suspend in the `acquire` step. If that is needed, I think users should prefer `onClose`, or `Resource` instead.

So I would prefer to revert this also.